src/anemoi/training/diagnostics/mlflow/logger.py

```diff
@@ -378,10 +378,59 @@ def experiment(self) -> MLFlowLogger.experiment:
     def log_system_metrics(self) -> None:
         """Log system metrics (CPU, GPU, etc)."""
         import mlflow
+        import psutil
+        from mlflow.system_metrics.metrics.base_metrics_monitor import BaseMetricsMonitor
+        from mlflow.system_metrics.metrics.disk_monitor import DiskMonitor
+        from mlflow.system_metrics.metrics.gpu_monitor import GPUMonitor
+        from mlflow.system_metrics.metrics.network_monitor import NetworkMonitor
         from mlflow.system_metrics.system_metrics_monitor import SystemMetricsMonitor
 
+        class CustomCPUMonitor(BaseMetricsMonitor):
+            """Class for monitoring CPU stats.
+
+            Extends default CPUMonitor, to also measure total \
+                    memory and a different formula for calculating used memory.
+
+            """
+
+            def collect_metrics(self) -> None:
+                # Get CPU metrics.
+                cpu_percent = psutil.cpu_percent()
+                self._metrics["cpu_utilization_percentage"].append(cpu_percent)
+
+                system_memory = psutil.virtual_memory()
+                # Change the formula for measuring CPU memory usage
+                # By default Mlflow uses psutil.virtual_memory().used
+                # Tests have shown that "used" underreports memory usage by as much as a factor of 2,
+                #   "used" also misses increased memory usage from using a higher prefetch factor
+                self._metrics["system_memory_usage_megabytes"].append(
+                    (system_memory.total - system_memory.available) / 1e6,
```

Comment:
Also, I'm not sure it does even capture other slurm users processes. I ran on a shared GPU node on Atos and the memory consumption reported using `self._metrics["system_memory_usage_megabytes"].append(
                    (system_memory.total - system_memory.available) / 1e6,` was firmly under my 1/4 slice

Reply:
`active` is the psutil's `active`. As everything else, "system dependent" :) For usecases where this run in isolated environments (k8s, slurm, etc), the system total / available could work as there is no inference, but it may massively surprise users who train on their laptops, run parallel trainings, etc

also note that linux by definition tries not to have (almost) any free memory -- it just tries to occupy the page cache as much as it can to speed disk operations (and its a bad idea to fight it). Similarly for libraries like numpy -- they just hold onto any memory allocated, even though they don't really need it (anymore)

the rss (or pss, though the difference is quite marg... the parallelizations here are usually share nothing, right) summing approach sounds the cleanest, but may not be worth it, depends on what the reporting is needed for

Reply:
I reran with `active` to see what it looks like, compared to "total-available". With 4 dataloaders per node and 16 dataloaders per node. At 20 dataloaders I run out of memory and the job crashes.
<br>

<img width="1483" alt="Screenshot 2024-10-17 at 13 20 26" src="https://github.com/user-attachments/assets/14b6c24e-668f-483f-98b0-97500523a6a6">

I don't know what `active` measures but it doesn't seem to reflect memory usage very well in this case (comparing the purple and brown lines).

I don't think we have to worry about people on laptops (yet!) but we can of course revisit in the future.

Reply:
yeah, no worries, I was just curious, dont know that much about how exactly this is intended to be used